File: legacy/python-backend/app/backtesting/metrics/performance_metrics.py
import numpy as np


class PerformanceMetrics:

    @staticmethod
    def calculate_total_return(
        initial_balance,
        final_balance
    ):

        return (
            (
                final_balance
                - initial_balance
            )
            / initial_balance
        ) * 100

    @staticmethod
    def calculate_sharpe_ratio(
        returns,
        risk_free_rate=0
    ):

        returns = np.array(returns)

        if returns.std() == 0:
            return 0

        return (
            (
                returns.mean()
                - risk_free_rate
            )
            / returns.std()
        )

    @staticmethod
    def calculate_max_drawdown(
        equity_curve
    ):

        peak = equity_curve[0]

        max_drawdown = 0

        for value in equity_curve:

            if value > peak:
                peak = value

            drawdown = (
                peak - value
            ) / peak

            if drawdown > max_drawdown:
                max_drawdown = drawdown

        return max_drawdown * 100

File: backend/src/backtesting/metrics/performanceMetrics.ts
export class PerformanceMetrics {
  static calculateTotalReturn(initialBalance: number, finalBalance: number): number {
    return ((finalBalance - initialBalance) / initialBalance) * 100;
  }

  static calculateSharpeRatio(returns: number[], riskFreeRate: number = 0): number {
    if (returns.length === 0) return 0;

    const mean = returns.reduce((sum, val) => sum + val, 0) / returns.length;

    const variance = returns.reduce((sum, val) => sum + Math.pow(val - mean, 2), 0) / returns.length;
    const stdDev = Math.sqrt(variance);

    if (stdDev === 0) return 0;

    return (mean - riskFreeRate) / stdDev;
  }

  static calculateMaxDrawdown(equityCurve: number[]): number {
    if (equityCurve.length === 0) return 0;

    let peak = equityCurve[0];
    let maxDrawdown = 0;

    for (const value of equityCurve) {
      if (value > peak) {
        peak = value;
      }

      const drawdown = (peak - value) / peak;
      if (drawdown > maxDrawdown) {
        maxDrawdown = drawdown;
      }
    }

    return maxDrawdown * 100;
  }
}
export default PerformanceMetrics;
